### DISCUSSION TITLE
Best way to do a MenuButton clear()

### DISCUSSION THREAD

**Discussion by TomDeRybel** [fltk-rs/fltk-rs#1305] [Q&A]
Hello,

I've got a MenuButton that I have to re-populate when the "control mode" changes (due to changing the selection in this MenuButton). 

Until now, I did a clear() and then an add_choice().
This worked, but as of a recent update, I now get errors saying this is unsafe (and this is reflected in the documentation).

My question is: what is a good way to do this, that is also safe (even if it requires an unsafe() block).

My naive approach would be to do this:
unsafe {
eq_ctl_widgets[stage_n].shape_type.clear();
eq_ctl_widgets[stage_n].shape_type.add_choice("Q|OBW");
}

But, that does not handle the warning in the documentation about string invalidation. I'm not using the string I got from the MenuButton after the clear() in the call-back, which is likely why this works :-). 

Though, some way to make sure I don't goof at a later stage would be a good thing. eg: 
- Some way to make that invalidated string an empty "" string, or something that gets caught during compilation?
- The docs mention "clearing while it is popped-up causes string invalidation". Is there a way to not make it popped-up before the clear()?

Anyhow, I'm clearly not sure how to go about this. Any and all hints are most welcome!

Best Regards,
Tom

**Comment by MoAlyousef** [comment]
Hello Tom

Well it's mostly safe. I wasn't able to pinpoint the problem exactly, even while using valgrind and sanitizers. The invalidated string is delivered to pango and this would cause issues. Unfortunately there is no way to find out whether a MenuButton is popped up AFAIK. If clearing happens in a callback nothing is triggered, so I would say it's safe. 

The problem occurs when the menu is popped up and something else clears it as can be demonstrated here:
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    let a = fltk::app::App::default();
    let mut w = window::Window::default().with_size(400, 300);
    let mut f = frame::Frame::new(0, 0, 200, 150, "Left click here!").center_of_parent();
    f.set_frame(FrameType::BorderFrame);
    f.set_color(Color::Red);
    let mut m = menu::MenuButton::default().size_of(&f).center_of(&f);
    m.set_type(menu::MenuButtonType::Popup3);
    m.add_choice("Q|OBW");
    w.end();
    w.show();
    app::add_timeout3(1.0, move |a| {
        unsafe {
            m.clear();
            m.add_choice("Q|OBW");
        }
        app::repeat_timeout3(1.0, a);
    });
    a.run().unwrap();
}
```

**Comment by TomDeRybel** [comment]
Thank you, again, for your fast and clear reply!

I'll go and document this in the affected file, in case future me discovers a new foot gun :-)

**Reply by MoAlyousef** [reply]
So I have inquired at the [FLTK mailing list](https://groups.google.com/g/fltkgeneral/c/eTyKlUMtGU8), and it appears that checking for grabbing windows would suffice. So I've added a check and removed the unsafe annotation.

**Reply by TomDeRybel** [reply]
Wow, thank you for looking into this so fast. I'll make sure to test this when the next fltk-rs is released. (I tried to test it just now, but got stuck in conflicting fltk-sys dependencies with fltk-grid.)

**Reply by TomDeRybel** [reply]
I forgot to reply, my apologies. This is to confirm that the changes you made work for my application.
